source/Attribute.cpp

```diff
@@ -0,0 +1,425 @@
+/* Attribute.cpp
+Copyright (c) 2023 by tibetiroka
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "Attribute.h"
+
+#include <algorithm>
+#include <limits>
+#include <unordered_map>
+
+using namespace std;
+
+const string Attribute::effectNames[] = {"shields", "hull", "thrust", "reverse thrust", "turn",
+		"cooling", "active cooling", "cloak", "force", "energy", "fuel", "heat", "discharge", "corrosion", "leak", "burn",
+		"ion", "scramble", "slowing", "disruption", "disabled", "minable", "ramscoop", "piercing", "delay", "depleted delay"};
+
+// Some category names are always the same as the corresponding effect names.
+const string Attribute::categoryNames[] = {"shield generation", "hull repair rate",
+		GetEffectName(THRUST), GetEffectName(REVERSE_THRUST), GetEffectName(TURN), GetEffectName(COOLING),
+		GetEffectName(ACTIVE_COOLING), GetEffectName(CLOAK), "afterburner thrust", "firing", "protection",
+		"resistance", "damage", "capacity"};
+
+bool Attribute::cached = false;
+
+
+
+// Creates a new categorized attribute. Use -1 if there is no category or effect in the definition.
+Attribute::Attribute(AttributeCategory category, AttributeEffect effect, AttributeEffect secondary, bool usePreferred)
+{
+	if(usePreferred)
```

Comment:
gotta be honest this confuses me a bit
is this for guessing what it is?

Reply:
~~you would know if you started with the enums~~
Some attributes can be described in multiple ways. This usually means shortcuts, like "shield generation" instead of "shield generation shields". These all have a preferred version that we use in the engine to avoid accidentally duplicating them. If `usePreferred` is enabled, it will make sure that the created attribute is the preferred version by tweaking it a bit. Otherwise, it just creates the attribute with the given things.

Reply:
can you add smth like that in a comment or smth?